packages/url_launcher/url_launcher_ios/ios/url_launcher_ios/Sources/url_launcher_ios/Launcher.swift

```diff
@@ -18,5 +18,21 @@ protocol Launcher {
     completionHandler completion: ((Bool) -> Void)?)
 }
 
-/// Launcher is intentionally a direct passthroguh to UIApplication.
-extension UIApplication: Launcher {}
+// TODO(hellohuanlin): This wrapper is a workaround for iOS 18 Beta 3 where completionHandler is annotated with @MainActor @Sendable, resulting in compile error when conforming UIApplication to Launcher. We should try again in newer betas.
+/// A default URL launcher.
+final class DefaultLauncher: Launcher {
```

Comment:
@hellohuanlin `Launcher` is documented to only exist for testing purposes.  
https://github.com/flutter/packages/blob/01527175e38e7ecd9ec180a1f0781580adb82dd5/packages/url_launcher/url_launcher_ios/ios/url_launcher_ios/Sources/url_launcher_ios/Launcher.swift#L7-L9

I know you [reviewed the PR](https://github.com/flutter/packages/pull/4753/files#diff-c38ad4a9a3711ebe154cfb06c02721450bbd4d80beedb7115b15c58f31d26deeR136), based on [your `DeviceProtocol` example](https://docs.google.com/document/d/1XsaulkJA6_ZSpM7chkQLhQY25sQqhEMiqHMYzw8H85o/edit?resourcekey=0-_cUjF1c0iBvRLKfV-3gK5A#heading=h.5t23pi6mihh3), that's the idiomatic way you suggest doing this, or is there another way?

Reply:
The idiomatic way is (as shown in wwdc's demo code a few years ago): 
  
(1) for real dependency, extend system type to conform to our protocol (e.g. `extension UIApplication: Launcher`)
(2) for fake dependency, define a custom type that conforms to the same protocol (e.g. `class MockLauncher: Launcher`)

In this case, (1) doesn't work anymore. So we have to define the custom type (e.g. `class DefaultLauncher: Launcher`). 

However, I highly suspect Apple may fix it in newer betas, so that we can revert this PR. As discussed in the PR description, Apple seems to suggest that if a module is not ready to adopt swift concurrency, it should just work as before. So they are not supposed to break us.

Reply:
looks like ios18 and xcode update break it...